AssertionError

Appears to be a regression in 0.9.0:

```
[error] java.lang.AssertionError: assertion failed: method overloads mismatch; bytecode=List(def minSize: (I)Lorg.scalatest.prop.Configuration$MinSize;, def minSize: ()Ljava.util.concurrent.atomic.AtomicReference;) pickle=Stream(SymbolInfo(VALsym, minSize, owner=155, isScopedPrivate=true)) for object org.scalatest.prop.Configuration
[error] 	at scala.Predef$.assert(Predef.scala:223)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$8(MimaUnpickler.scala:66)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$8$adapted(MimaUnpickler.scala:63)
[error] 	at scala.collection.immutable.Map$Map4.foreach(Map.scala:492)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.doMethods$1(MimaUnpickler.scala:63)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$18(MimaUnpickler.scala:91)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$18$adapted(MimaUnpickler.scala:86)
```

typelevel/cats-testkit-scalatest#144

--- a/core/src/main/scala/com/typesafe/tools/mima/core/ClassfileParser.scala
+++ b/core/src/main/scala/com/typesafe/tools/mima/core/ClassfileParser.scala
@@ -133,8 +133,8 @@ final class ClassfileParser private (in: BufferReader, pool: ConstantPool) {
 
     val numAnnots = in.nextChar
     var i = 0
-    var bytes: Array[Byte] = null
-    while (i < numAnnots && bytes == null) {
+    var bytes = new Array[Byte](0)
+    while (i < numAnnots && bytes.length == 0) {
       pool.getExternalName(in.nextChar) match {
         case ScalaSignatureAnnot     => checkScalaSigAnnotArg(); bytes = parseScalaSigBytes()
         case ScalaLongSignatureAnnot => checkScalaSigAnnotArg(); bytes = parseScalaLongSigBytes()
@@ -142,10 +142,7 @@ final class ClassfileParser private (in: BufferReader, pool: ConstantPool) {
       }
       i += 1
     }
-    if (bytes != null) {
-      val pb = new PickleBuffer(bytes)
-      MimaUnpickler.unpickleClass(pb, clazz, in.path)
-    }
+    MimaUnpickler.unpickleClass(new PickleBuffer(bytes), clazz, in.path)
   }
 
   private final val ScalaSignatureAnnot     = "Lscala.reflect.ScalaSignature;"

--- a/core/src/main/scala/com/typesafe/tools/mima/core/MimaUnpickler.scala
+++ b/core/src/main/scala/com/typesafe/tools/mima/core/MimaUnpickler.scala
@@ -3,7 +3,9 @@ package com.typesafe.tools.mima.core
 import PickleFormat._
 
 object MimaUnpickler {
-  def unpickleClass(buf: PickleBuffer, clazz: ClassInfo, path: String) = {
+  def unpickleClass(buf: PickleBuffer, clazz: ClassInfo, path: String): Unit = {
+    if (buf.bytes.length == 0) return
+
     buf.readNat(); buf.readNat() // major, minor version
 
     val index    = buf.createIndex
@@ -18,7 +20,7 @@ object MimaUnpickler {
 
     // SymbolInfo = name_Ref owner_Ref flags_LongNat [privateWithin_Ref] info_Ref
     def readSymbol(): SymbolInfo = {
-      val tag   = buf.lastByte()
+      val tag   = buf.readByte()
       val end   = buf.readNat() + buf.readIndex
       val name  = entries.nameAt(buf.readNat())
       val owner = buf.readNat()
@@ -32,7 +34,7 @@ object MimaUnpickler {
     def symbolToClass(symbolInfo: SymbolInfo) = {
       if (symbolInfo.name == REFINE_CLASS_NAME) {
         // eg: CLASSsym 4: 89(<refinement>) 0 0[] 87
-        // The UnPickler in the compiler also excludes these with "isRefinementSymbolEntry"
+        // Nsc's UnPickler also excludes these with "isRefinementSymbolEntry"
         NoClass
       } else if (symbolInfo.name == "<local child>") {
         // Predef$$less$colon$less$<local child>
@@ -60,12 +62,21 @@ object MimaUnpickler {
       methods.iterator
         .filter(!_.isParam)
         .filter(_.name != "<init>") // TODO support package private constructors
-        .toSeq.groupBy(_.name).foreach { case (name, overloads) =>
-          val methods = clazz.methods.get(name).filter(!_.isBridge).toList
-          if (methods.nonEmpty && overloads.exists(_.isScopedPrivate)) {
-            assert(overloads.size == methods.size, s"method overloads mismatch; bytecode=$methods pickle=$overloads for ${clazz.description}")
-            methods.zip(overloads).foreach { case (method, symbolInfo) =>
-              method.scopedPrivate = symbolInfo.isScopedPrivate
+        .toSeq.groupBy(_.name).foreach { case (name, pickleMethods) =>
+          val bytecodeMethods = clazz.methods.get(name).filter(!_.isBridge).toList
+          // #630 one way this happens with mixins:
+          //    trait Foo { def bar(x: Int): Int = x }
+          //    class Bar extends Foo { private[foo] def bar: String = "" }
+          // during pickling Bar only contains the package private bar()String
+          // but later in the backend the classfile gets a copy of bar(Int)Int
+          // so the "bar" method in the pickle doesn't know which bytecode method it's about
+          // the proper way to fix this involves unpickling the types in the pickle,
+          // then implementing the rules of erasure, so that you can then match the pickle
+          // types with the erased types.  Meanwhile we'll just ignore them, worst case users
+          // need to add a filter like they have for years.
+          if (pickleMethods.size == bytecodeMethods.size && pickleMethods.exists(_.isScopedPrivate)) {
+            bytecodeMethods.zip(pickleMethods).foreach { case (bytecodeMeth, pickleMeth) =>
+              bytecodeMeth.scopedPrivate = pickleMeth.isScopedPrivate
             }
           }
       }
@@ -73,10 +84,15 @@ object MimaUnpickler {
 
 
     for (num <- index.indices) {
-      buf.runAtIndex(index(num)) {
+      buf.atIndex(index(num)) {
         val tag = buf.readByte()
-        if (tag == CLASSsym || tag == MODULEsym || tag == VALsym)
-          syms(num) = readSymbol()
+        buf.readIndex -= 1
+        tag match {
+          case  CLASSsym => syms(num) = readSymbol()
+          case MODULEsym => syms(num) = readSymbol()
+          case    VALsym => syms(num) = readSymbol()
+          case _         =>
+        }
       }
     }
 
@@ -87,18 +103,15 @@ object MimaUnpickler {
       val clazz = classes(num)
       if (clsSym.isScopedPrivate)
         clazz.module._scopedPrivate = true
-      val methods = methSyms.collect { case (sym, _) if sym.owner == num => sym }
-      doMethods(clazz, methods.toList)
+      val methods = methSyms.collect { case (sym, _) if sym.owner == num => sym }.toList
+      doMethods(clazz, methods)
     }
   }
 
   final case class SymbolInfo(tag: Int, name: String, owner: Int, flags: Long, isScopedPrivate: Boolean) {
     def hasFlag(flag: Long): Boolean = (flags & flag) != 0L
     def isModuleOrModuleClass        = hasFlag(Flags.MODULE_PKL)
-    def isModuleClass                = isModuleOrModuleClass && tag == CLASSsym
-    def isModule                     = isModuleOrModuleClass && tag == MODULEsym
     def isParam                      = hasFlag(Flags.PARAM)
-    def mcmc                         = if (isModuleClass) "MC" else if (isModule) "M" else "C"
     override def toString = s"SymbolInfo(${tag2string(tag)}, $name, owner=$owner, isScopedPrivate=$isScopedPrivate)"
   }
 

--- a/core/src/main/scala/com/typesafe/tools/mima/core/PickleBuffer.scala
+++ b/core/src/main/scala/com/typesafe/tools/mima/core/PickleBuffer.scala
@@ -4,7 +4,6 @@ final class PickleBuffer(val bytes: Array[Byte]) {
   var readIndex = 0
 
   def readByte(): Int = { val x = bytes(readIndex).toInt; readIndex += 1; x }
-  def lastByte(): Int = bytes(readIndex - 1).toInt
 
   /** Read a natural number in big endian format, base 128. All but the last digits have bit 0x80 set. */
   def readNat(): Int = readLongNat().toInt
@@ -32,11 +31,11 @@ final class PickleBuffer(val bytes: Array[Byte]) {
   }
 
   def createIndex: Array[Int] = {
-    runAtIndex(0) {
+    atIndex(0) {
       readNat(); readNat() // discard version
       0.until(readNat()).iterator.map { _ =>
         val idx = readIndex
-        val tag = readByte()
+        readByte() // tag
         val len = readNat()
         readIndex += len
         idx
@@ -49,7 +48,7 @@ final class PickleBuffer(val bytes: Array[Byte]) {
    */
   def toIndexedSeq: IndexedSeq[(Int, Array[Byte])] = {
     for (idx <- createIndex) yield {
-      runAtIndex(idx) {
+      atIndex(idx) {
         val tag = readByte()
         val len = readNat()
         tag -> bytes.slice(readIndex, readIndex + len)
@@ -57,7 +56,7 @@ final class PickleBuffer(val bytes: Array[Byte]) {
     }
   }
 
-  def runAtIndex[T](i: Int)(body: => T): T = {
+  def atIndex[T](i: Int)(body: => T): T = {
     val saved = readIndex
     readIndex = i
     try body finally readIndex = saved

--- a/core/src/main/scala/com/typesafe/tools/mima/core/PickleFormat.scala
+++ b/core/src/main/scala/com/typesafe/tools/mima/core/PickleFormat.scala
@@ -6,104 +6,6 @@ object PickleFormat {
    * Symbol table attribute format:
    *   Symtab         = nentries_Nat {Entry}
    *
-   *   Entry          = 1 TERMNAME len_Nat NameInfo
-   *                  | 2 TYPENAME len_Nat NameInfo
-   *                  |
-   *                  | 3   NONEsym len_Nat
-   *                  | 4   TYPEsym len_Nat SymbolInfo
-   *                  | 5  ALIASsym len_Nat SymbolInfo
-   *                  | 6  CLASSsym len_Nat SymbolInfo [thistype_Ref]
-   *                  | 7 MODULEsym len_Nat SymbolInfo
-   *                  | 8    VALsym len_Nat [defaultGetter_Ref] SymbolInfo [alias_Ref]  // defaultGetter_Ref no longer needed
-   *                  |
-   *                  | 9          EXTref len_Nat name_Ref [owner_Ref]
-   *                  | 10 EXTMODCLASSref len_Nat name_Ref [owner_Ref]
-   *                  |
-   *                  | 11             NOtpe len_Nat
-   *                  | 12       NOPREFIXtpe len_Nat
-   *                  | 13           THIStpe len_Nat      sym_Ref
-   *                  | 14         SINGLEtpe len_Nat     type_Ref  sym_Ref
-   *                  | 15       CONSTANTtpe len_Nat constant_Ref
-   *                  | 16        TYPEREFtpe len_Nat     type_Ref  sym_Ref {targ_Ref}
-   *                  | 17     TYPEBOUNDStpe len_Nat      tpe_Ref  tpe_Ref
-   *                  | 18        REFINEDtpe len_Nat classsym_Ref {tpe_Ref}
-   *                  | 19      CLASSINFOtpe len_Nat classsym_Ref {tpe_Ref}
-   *                  | 20         METHODtpe len_Nat      tpe_Ref {sym_Ref}
-   *                  | 21          POLYTtpe len_Nat      tpe_Ref {sym_Ref}
-   *                  | 22 IMPLICITMETHODtpe len_Nat      tpe_Ref {sym_Ref}   // deprecated
-   *                  | 52          SUPERtpe len_Nat      tpe_Ref  tpe_Ref
-   *                  |
-   *                  | 24 LITERALunit    len_Nat
-   *                  | 25 LITERALboolean len_Nat value_Long
-   *                  | 26 LITERALbyte    len_Nat value_Long
-   *                  | 27 LITERALshort   len_Nat value_Long
-   *                  | 28 LITERALchar    len_Nat value_Long
-   *                  | 29 LITERALint     len_Nat value_Long
-   *                  | 30 LITERALlong    len_Nat value_Long
-   *                  | 31 LITERALfloat   len_Nat value_Long
-   *                  | 32 LITERALdouble  len_Nat value_Long
-   *                  | 33 LITERALstring  len_Nat  name_Ref
-   *                  | 34 LITERALnull    len_Nat
-   *                  | 35 LITERALclass   len_Nat   tpe_Ref
-   *                  | 36 LITERALenum    len_Nat   sym_Ref
-   *                  | 37 LITERALsymbol  len_Nat  name_Ref
-   *                  |
-   *                  | 40 SYMANNOT         len_Nat            sym_Ref  AnnotInfoBody
-   *                  | 41 CHILDREN         len_Nat            sym_Ref     {sym_Ref}
-   *                  | 42     ANNOTATEDtpe len_Nat           [sym_Ref]     tpe_Ref {annotinfo_Ref} // sym_Ref no longer needed
-   *                  | 43 ANNOTINFO        len_Nat      AnnotInfoBody
-   *                  | 44 ANNOTARGARRAY    len_Nat {constAnnotArg_Ref}
-   *                  | 47 DEBRUIJNINDEXtpe len_Nat          level_Nat    index_Nat   // deprecated
-   *                  | 48   EXISTENTIALtpe len_Nat           type_Ref  {symbol_Ref}
-   *                  |
-   *                  | 49 TREE len_Nat
-   *                  |    1           EMPTYtree
-   *                  |    2         PACKAGEtree type_Ref      sym_Ref        mods_Ref  name_Ref      {tree_Ref}
-   *                  |    3           CLASStree type_Ref      sym_Ref        mods_Ref  name_Ref       tree_Ref {tree_Ref}
-   *                  |    4          MODULEtree type_Ref      sym_Ref        mods_Ref  name_Ref       tree_Ref
-   *                  |    5          VALDEFtree type_Ref      sym_Ref        mods_Ref  name_Ref       tree_Ref  tree_Ref
-   *                  |    6          DEFDEFtree type_Ref      sym_Ref        mods_Ref  name_Ref numtparams_Nat {tree_Ref} numparamss_Nat {numparams_Nat {tree_Ref}} tree_Ref tree_Ref
-   *                  |    7         TYPEDEFtree type_Ref      sym_Ref        mods_Ref  name_Ref       tree_Ref {tree_Ref}
-   *                  |    8           LABELtree type_Ref      sym_Ref        tree_Ref {tree_Ref}
-   *                  |    9          IMPORTtree type_Ref      sym_Ref        tree_Ref {name_Ref       name_Ref}
-   *                  |   11          DOCDEFtree type_Ref      sym_Ref      string_Ref  tree_Ref
-   *                  |   12        TEMPLATEtree type_Ref      sym_Ref  numparents_Nat {tree_Ref}      tree_Ref {tree_Ref}
-   *                  |   13           BLOCKtree type_Ref     tree_Ref       {tree_Ref}
-   *                  |   14            CASEtree type_Ref     tree_Ref        tree_Ref  tree_Ref
-   *                  |   15        SEQUENCEtree type_Ref    {tree_Ref}
-   *                  |   16     ALTERNATIVEtree type_Ref    {tree_Ref}
-   *                  |   17            STARtree type_Ref    {tree_Ref}
-   *                  |   18            BINDtree type_Ref      sym_Ref  name_Ref  tree_Ref
-   *                  |   19         UNAPPLYtree type_Ref     tree_Ref {tree_Ref}
-   *                  |   20      ARRAYVALUEtree type_Ref     tree_Ref {tree_Ref}
-   *                  |   21        FUNCTIONtree type_Ref      sym_Ref  tree_Ref {tree_Ref}
-   *                  |   22          ASSIGNtree type_Ref     tree_Ref  tree_Ref
-   *                  |   23              IFtree type_Ref     tree_Ref  tree_Ref  tree_Ref
-   *                  |   24           MATCHtree type_Ref     tree_Ref {tree_Ref}
-   *                  |   25          RETURNtree type_Ref      sym_Ref  tree_Ref
-   *                  |   26             TREtree type_Ref     tree_Ref  tree_Ref {tree_Ref}
-   *                  |   27           THROWtree type_Ref     tree_Ref
-   *                  |   28             NEWtree type_Ref     tree_Ref
-   *                  |   29           TYPEDtree type_Ref     tree_Ref  tree_Ref
-   *                  |   30       TYPEAPPLYtree type_Ref     tree_Ref {tree_Ref}
-   *                  |   31           APPLYtree type_Ref     tree_Ref {tree_Ref}
-   *                  |   32    APPLYDYNAMICtree type_Ref      sym_Ref  tree_Ref {tree_Ref}
-   *                  |   33           SUPERtree type_Ref      sym_Ref  tree_Ref  name_Ref
-   *                  |   34            THIStree type_Ref      sym_Ref  name_Ref
-   *                  |   35          SELECTtree type_Ref      sym_Ref  tree_Ref  name_Ref
-   *                  |   36           IDENTtree type_Ref      sym_Ref  name_Ref
-   *                  |   37         LITERALtree type_Ref constant_Ref
-   *                  |   38            TYPEtree type_Ref
-   *                  |   39       ANNOTATEDtree type_Ref     tree_Ref  tree_Ref
-   *                  |   40   SINGLETONTYPEtree type_Ref     tree_Ref
-   *                  |   41  SELECTFROMTYPEtree type_Ref     tree_Ref  name_Ref
-   *                  |   42    COMPOUNDTYPEtree type_Ref     tree_Ref
-   *                  |   43     APPLIEDTYPEtree type_Ref     tree_Ref {tree_Ref}
-   *                  |   44      TYPEBOUNDStree type_Ref     tree_Ref  tree_Ref
-   *                  |   45 EXISTENTIALTYPEtree type_Ref     tree_Ref {tree_Ref}
-   *                  |
-   *                  | 50 MODIFIERS len_Nat flags_Long privateWithin_Ref
-   *
    *         NameInfo     = <character sequence of length len_Nat in UTF-8 format>
    *       SymbolInfo     = name_Ref owner_Ref flags_LongNat [privateWithin_Ref] info_Ref
    *             Long     = <len_Nat-byte signed number in big endian format>
@@ -113,6 +15,105 @@ object PickleFormat {
    *   ConstAnnotArg      = Constant | AnnotInfo | AnnotArgArray
    *
    *   len is remaining length after `len`.
+   *
+   *   Entry = \
+   *     1 TERMNAME len_Nat NameInfo
+   *   | 2 TYPENAME len_Nat NameInfo
+   *   |
+   *   | 3   NONEsym len_Nat
+   *   | 4   TYPEsym len_Nat SymbolInfo
+   *   | 5  ALIASsym len_Nat SymbolInfo
+   *   | 6  CLASSsym len_Nat SymbolInfo [thistype_Ref]
+   *   | 7 MODULEsym len_Nat SymbolInfo
+   *   | 8    VALsym len_Nat [defaultGetter_Ref] SymbolInfo [alias_Ref]  // defaultGetter_Ref no longer needed
+   *   |
+   *   | 9          EXTref len_Nat name_Ref [owner_Ref]
+   *   | 10 EXTMODCLASSref len_Nat name_Ref [owner_Ref]
+   *   |
+   *   | 11             NOtpe len_Nat
+   *   | 12       NOPREFIXtpe len_Nat
+   *   | 13           THIStpe len_Nat      sym_Ref
+   *   | 14         SINGLEtpe len_Nat     type_Ref  sym_Ref
+   *   | 15       CONSTANTtpe len_Nat constant_Ref
+   *   | 16        TYPEREFtpe len_Nat     type_Ref  sym_Ref {targ_Ref}
+   *   | 17     TYPEBOUNDStpe len_Nat      tpe_Ref  tpe_Ref
+   *   | 18        REFINEDtpe len_Nat classsym_Ref {tpe_Ref}
+   *   | 19      CLASSINFOtpe len_Nat classsym_Ref {tpe_Ref}
+   *   | 20         METHODtpe len_Nat      tpe_Ref {sym_Ref}
+   *   | 21          POLYTtpe len_Nat      tpe_Ref {sym_Ref}
+   *   | 22 IMPLICITMETHODtpe len_Nat      tpe_Ref {sym_Ref}   // deprecated
+   *   | 52          SUPERtpe len_Nat      tpe_Ref  tpe_Ref
+   *   |
+   *   | 24 LITERALunit    len_Nat
+   *   | 25 LITERALboolean len_Nat value_Long
+   *   | 26 LITERALbyte    len_Nat value_Long
+   *   | 27 LITERALshort   len_Nat value_Long
+   *   | 28 LITERALchar    len_Nat value_Long
+   *   | 29 LITERALint     len_Nat value_Long
+   *   | 30 LITERALlong    len_Nat value_Long
+   *   | 31 LITERALfloat   len_Nat value_Long
+   *   | 32 LITERALdouble  len_Nat value_Long
+   *   | 33 LITERALstring  len_Nat  name_Ref
+   *   | 34 LITERALnull    len_Nat
+   *   | 35 LITERALclass   len_Nat   tpe_Ref
+   *   | 36 LITERALenum    len_Nat   sym_Ref
+   *   | 37 LITERALsymbol  len_Nat  name_Ref
+   *   |
+   *   | 40 SYMANNOT         len_Nat            sym_Ref  AnnotInfoBody
+   *   | 41 CHILDREN         len_Nat            sym_Ref     {sym_Ref}
+   *   | 42     ANNOTATEDtpe len_Nat           [sym_Ref]     tpe_Ref {annotinfo_Ref} // sym_Ref no longer needed
+   *   | 43 ANNOTINFO        len_Nat      AnnotInfoBody
+   *   | 44 ANNOTARGARRAY    len_Nat {constAnnotArg_Ref}
+   *   | 47 DEBRUIJNINDEXtpe len_Nat          level_Nat    index_Nat   // deprecated
+   *   | 48   EXISTENTIALtpe len_Nat           type_Ref  {symbol_Ref}
+   *   |
+   *   | 49 TREE len_Nat
+   *   |    1           EMPTYtree
+   *   |    2         PACKAGEtree type_Ref      sym_Ref        mods_Ref  name_Ref      {tree_Ref}
+   *   |    3           CLASStree type_Ref      sym_Ref        mods_Ref  name_Ref       tree_Ref {tree_Ref}
+   *   |    4          MODULEtree type_Ref      sym_Ref        mods_Ref  name_Ref       tree_Ref
+   *   |    5          VALDEFtree type_Ref      sym_Ref        mods_Ref  name_Ref       tree_Ref  tree_Ref
+   *   |    6          DEFDEFtree type_Ref      sym_Ref        mods_Ref  name_Ref numtparams_Nat {tree_Ref} numparamss_Nat {numparams_Nat {tree_Ref}} tree_Ref tree_Ref
+   *   |    7         TYPEDEFtree type_Ref      sym_Ref        mods_Ref  name_Ref       tree_Ref {tree_Ref}
+   *   |    8           LABELtree type_Ref      sym_Ref        tree_Ref {tree_Ref}
+   *   |    9          IMPORTtree type_Ref      sym_Ref        tree_Ref {name_Ref       name_Ref}
+   *   |   11          DOCDEFtree type_Ref      sym_Ref      string_Ref  tree_Ref
+   *   |   12        TEMPLATEtree type_Ref      sym_Ref  numparents_Nat {tree_Ref}      tree_Ref {tree_Ref}
+   *   |   13           BLOCKtree type_Ref     tree_Ref       {tree_Ref}
+   *   |   14            CASEtree type_Ref     tree_Ref        tree_Ref  tree_Ref
+   *   |   15        SEQUENCEtree type_Ref    {tree_Ref}
+   *   |   16     ALTERNATIVEtree type_Ref    {tree_Ref}
+   *   |   17            STARtree type_Ref    {tree_Ref}
+   *   |   18            BINDtree type_Ref      sym_Ref  name_Ref  tree_Ref
+   *   |   19         UNAPPLYtree type_Ref     tree_Ref {tree_Ref}
+   *   |   20      ARRAYVALUEtree type_Ref     tree_Ref {tree_Ref}
+   *   |   21        FUNCTIONtree type_Ref      sym_Ref  tree_Ref {tree_Ref}
+   *   |   22          ASSIGNtree type_Ref     tree_Ref  tree_Ref
+   *   |   23              IFtree type_Ref     tree_Ref  tree_Ref  tree_Ref
+   *   |   24           MATCHtree type_Ref     tree_Ref {tree_Ref}
+   *   |   25          RETURNtree type_Ref      sym_Ref  tree_Ref
+   *   |   26             TREtree type_Ref     tree_Ref  tree_Ref {tree_Ref}
+   *   |   27           THROWtree type_Ref     tree_Ref
+   *   |   28             NEWtree type_Ref     tree_Ref
+   *   |   29           TYPEDtree type_Ref     tree_Ref  tree_Ref
+   *   |   30       TYPEAPPLYtree type_Ref     tree_Ref {tree_Ref}
+   *   |   31           APPLYtree type_Ref     tree_Ref {tree_Ref}
+   *   |   32    APPLYDYNAMICtree type_Ref      sym_Ref  tree_Ref {tree_Ref}
+   *   |   33           SUPERtree type_Ref      sym_Ref  tree_Ref  name_Ref
+   *   |   34            THIStree type_Ref      sym_Ref  name_Ref
+   *   |   35          SELECTtree type_Ref      sym_Ref  tree_Ref  name_Ref
+   *   |   36           IDENTtree type_Ref      sym_Ref  name_Ref
+   *   |   37         LITERALtree type_Ref constant_Ref
+   *   |   38            TYPEtree type_Ref
+   *   |   39       ANNOTATEDtree type_Ref     tree_Ref  tree_Ref
+   *   |   40   SINGLETONTYPEtree type_Ref     tree_Ref
+   *   |   41  SELECTFROMTYPEtree type_Ref     tree_Ref  name_Ref
+   *   |   42    COMPOUNDTYPEtree type_Ref     tree_Ref
+   *   |   43     APPLIEDTYPEtree type_Ref     tree_Ref {tree_Ref}
+   *   |   44      TYPEBOUNDStree type_Ref     tree_Ref  tree_Ref
+   *   |   45 EXISTENTIALTYPEtree type_Ref     tree_Ref {tree_Ref}
+   *   |
+   *   | 50 MODIFIERS len_Nat flags_Long privateWithin_Ref
    */
 
   val MajorVersion = 5
@@ -221,6 +222,10 @@ object PickleFormat {
     final val EXISTENTIALTYPEtree = 45
 
   final val MODIFIERS         = 50
+
+  final val   firstSymTag = NONEsym
+  final val    lastSymTag = VALsym
+  final val lastExtSymTag = EXTMODCLASSref
 
   def tag2string(tag: Int): String = tag match {
     case TERMname          => "TERMname"

--- a/project/MimaSettings.scala
+++ b/project/MimaSettings.scala
@@ -24,6 +24,8 @@ object MimaSettings {
       // * com.typesafe.tools.mima.core.ProblemFilters
       // * com.typesafe.tools.mima.core.*Problem
       // * com.typesafe.tools.mima.core.util.log.Logging
+      exclude[Problem]("*mima.core.MimaUnpickler*"),
+      exclude[Problem]("*mima.core.PickleBuffer*"),
     ),
   )
 }

